table/tables/partition_test.go

```diff
@@ -700,6 +702,1514 @@ func TestIssue31629(t *testing.T) {
+	result.CheckContain("partition:p2")
+	result.CheckNotContainMore([]string{"partition:p0", "partition:p1", "partition:p3"})
+
+	tk.MustExec("CREATE TABLE tkey12_2 (col1 INT, col2 INT ,col3 INT ,col4 INT , UNIQUE KEY(col2, col3)" +
+		") PARTITION BY KEY(col2, col3) PARTITIONS 4")
+	tk.MustExec("INSERT INTO tkey12_2 values(20,1,1,1),(1,2,NULL,2),(3,3,3,3),(3,3,NULL,3),(4,4,4,4)," +
+		"(5,5,5,5), (6,6,null,6),(7,7,7,7),(8,8,8,8),(9,9,9,9),(10,10,10,5),(11,11,11,6),(12,12,12,12)," +
+		"(13,13,13,13),(14,14,null,14)")
+	result = tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 = 2 and col3 IS NULL")
+	result.CheckNotContainMore([]string{"partition:p1", "partition:p2", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 = 2 and col3 IS NULL").Check(testkit.Rows("1 2 <nil> 2"))
+	result = tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 = 2")
+	result.CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 = 2").Check(testkit.Rows("1 2 <nil> 2"))
+	tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 = 2").CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 IS NULL")
+	tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 IS NULL").CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	// Get the partition information from information_schema.partitions
+	result = tk.MustQuery("select PARTITION_NAME,PARTITION_ORDINAL_POSITION,PARTITION_METHOD,PARTITION_EXPRESSION " +
+		"FROM information_schema.partitions where TABLE_NAME = 'tkey12_2'")
+	result.Check(testkit.Rows("p0 1 KEY `col2`,`col3`", "p1 2 KEY `col2`,`col3`", "p2 3 KEY `col2`,`col3`", "p3 4 KEY `col2`,`col3`"))
+
+	// This tests caculating the boundary partition ID when it prunes partition table
+	tk.MustExec("create table tkey16 (a int) partition by key (a) partitions 12")
+	tk.MustExec("insert into tkey16 values (0), (1), (2), (3)")
+	tk.MustExec("insert into tkey16 select a + 4 from tkey16")
+	tk.MustExec("insert into tkey16 select a + 8 from tkey16")
+	tk.MustExec("select * from information_schema.partitions where partition_name is not null")
+}
+
+func TestKeyPartitionWithGeneralci(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("create database partitiondb4")
+	defer tk.MustExec("drop database partitiondb4")
+	tk.MustExec("use partitiondb4")
+	tk.MustExec("CREATE TABLE tkey6_1 (" +
+		"col1 INT NOT NULL," +
+		"col2 DATE NOT NULL," +
+		"col3 VARCHAR(12) NOT NULL," +
+		"col4 INT NOT NULL," +
+		"UNIQUE KEY (col3)" +
+		") CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci " +
+		"PARTITION BY KEY(col3) " +
+		"PARTITIONS 4")
+	err := tk.ExecToErr("INSERT INTO tkey6_1 VALUES(1, SYSDATE(), 'linpin', 1), (1, SYSDATE(), 'linpin', 5)")
```

Comment:
```suggestion
		"PARTITIONS 3")
	err := tk.ExecToErr("INSERT INTO tkey6_1 VALUES(1, SYSDATE(), 'Linpin', 1), (1, SYSDATE(), 'linpin', 5)")
```
Linpin will compare as equal to linpin for case insensitive collation, this works with non partitioned table and should also work with key partitioned tables. Note that the duplicate key is only checked by the underlying table/partition, so the mapping from row/value to partition must be collation aware, like using collate.Key() function for calculating the partition id.